File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/component/ComponentDependency.java
package ru.tinkoff.kora.kora.app.annotation.processor.component;

import com.squareup.javapoet.CodeBlock;
import ru.tinkoff.kora.annotation.processor.common.CommonClassNames;
import ru.tinkoff.kora.application.graph.TypeRef;
import ru.tinkoff.kora.kora.app.annotation.processor.GraphResolutionHelper;
import ru.tinkoff.kora.kora.app.annotation.processor.ProcessingContext;

import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.TypeMirror;
import javax.lang.model.util.Types;
import java.util.List;
import java.util.Set;

public sealed interface ComponentDependency {
    DependencyClaim claim();

    CodeBlock write(ProcessingContext ctx, List<ResolvedComponent> resolvedComponents, Set<Integer> promisedComponents);

    sealed interface SingleDependency extends ComponentDependency {
        ResolvedComponent component();
    }

    record TargetDependency(DependencyClaim claim, ResolvedComponent component) implements SingleDependency {

        @Override
        public CodeBlock write(ProcessingContext ctx, List<ResolvedComponent> resolvedComponents, Set<Integer> promisedComponents) {
            var node = promisedComponents.contains(component.index()) ? CodeBlock.of("$L.get()", this.component.name()) : CodeBlock.of("$L", this.component.name());
            return CodeBlock.of("g.get($L)", node);
        }
    }

    record WrappedTargetDependency(DependencyClaim claim, ResolvedComponent component) implements SingleDependency {

        @Override
        public CodeBlock write(ProcessingContext ctx, List<ResolvedComponent> resolvedComponents, Set<Integer> promisedComponents) {
            var node = promisedComponents.contains(component.index()) ? CodeBlock.of("$L.get()", this.component.name()) : CodeBlock.of("$L", this.component.name());
            return CodeBlock.of("g.get($L).value()", node);
        }
    }

    record NullDependency(DependencyClaim claim) implements ComponentDependency {
        @Override
        public CodeBlock write(ProcessingContext ctx, List<ResolvedComponent> resolvedComponents, Set<Integer> promisedComponents) {
            return switch (this.claim.claimType()) {
                case ONE_NULLABLE -> CodeBlock.of("($T) null", this.claim.type());
                case NULLABLE_VALUE_OF -> CodeBlock.of("($T<$T>) null", CommonClassNames.valueOf, this.claim.type());
                case NULLABLE_PROMISE_OF -> CodeBlock.of("($T<$T>) null", CommonClassNames.promiseOf, this.claim.type());
                default -> throw new IllegalArgumentException(this.claim.claimType().toString());
            };
        }
    }

    record ValueOfDependency(DependencyClaim claim, SingleDependency delegate) implements SingleDependency {
        @Override
        public CodeBlock write(ProcessingContext ctx, List<ResolvedComponent> resolvedComponents, Set<Integer> promisedComponents) {
            var node = promisedComponents.contains(delegate.component().index()) ? CodeBlock.of("$L.get()", delegate.component().name()) : CodeBlock.of("$L", delegate.component().name());
            if (this.delegate instanceof WrappedTargetDependency) {
                return CodeBlock.of("g.valueOf($L).map($T::value).map(v -> ($T) v)", node, CommonClassNames.wrapped, claim.type());
            }
            return CodeBlock.of("g.valueOf($L).map(v -> ($T) v)", node, claim.type());
        }

        @Override
        public ResolvedComponent component() {
            return delegate.component();
        }
    }

    record PromiseOfDependency(DependencyClaim claim, SingleDependency delegate) implements SingleDependency {
        @Override
        public CodeBlock write(ProcessingContext ctx, List<ResolvedComponent> resolvedComponents, Set<Integer> promisedComponents) {
            if (this.delegate instanceof WrappedTargetDependency) {
                return CodeBlock.of("g.promiseOf($L.get()).map($T::value).map(v -> ($T) v)", this.delegate.component().name(), CommonClassNames.wrapped, this.claim.type());
            }
            return CodeBlock.of("g.promiseOf($L.get()).map(v -> ($T) v)", delegate.component().name(), this.claim.type());
        }

        @Override
        public ResolvedComponent component() {
            return delegate.component();
        }
    }

    record TypeOfDependency(DependencyClaim claim) implements SingleDependency {
        @Override
        public CodeBlock write(ProcessingContext ctx, List<ResolvedComponent> resolvedComponents, Set<Integer> promisedComponents) {
            return this.buildTypeRef(ctx.types, this.claim.type());
        }

        private CodeBlock buildTypeRef(Types types, TypeMirror typeRef) {
            if (typeRef instanceof DeclaredType) {
                var b = CodeBlock.builder();
                var typeArguments = ((DeclaredType) typeRef).getTypeArguments();

                if (typeArguments.isEmpty()) {
                    b.add("$T.of($T.class)", TypeRef.class, types.erasure(typeRef));
                } else {
                    b.add("$T.<$T>of($T.class", TypeRef.class, typeRef, types.erasure(typeRef));
                    for (var typeArgument : typeArguments) {
                        b.add("$>,\n$L$<", buildTypeRef(types, typeArgument));
                    }
                    b.add("\n)");
                }
                return b.build();
            } else {
                return CodeBlock.of("$T.of($T.class)", TypeRef.class, typeRef);
            }
        }

        @Override
        public ResolvedComponent component() {
            return null;
        }
    }


    // AllOf dependencies has no resolved component: we will resolve them after graph building

    record AllOfDependency(DependencyClaim claim) implements ComponentDependency {
        @Override
        public CodeBlock write(ProcessingContext ctx, List<ResolvedComponent> resolvedComponents, Set<Integer> promisedComponents) {
            var codeBlock = CodeBlock.builder().add("$T.of(", CommonClassNames.all);
            var dependencies = GraphResolutionHelper.findDependenciesForAllOf(ctx, this.claim, resolvedComponents);
            for (int i = 0; i < dependencies.size(); i++) {
                var dependency = dependencies.get(i);
                if (i == 0) {
                    codeBlock.indent().add("\n");
                }
                codeBlock.add(dependency.write(ctx, resolvedComponents, promisedComponents));
                if (i == dependencies.size() - 1) {
                    codeBlock.unindent();
                } else {
                    codeBlock.add(",");
                }
                codeBlock.add("\n");
            }

            return codeBlock.add("  )").build();
        }
    }

    record PromisedProxyParameterDependency(ru.tinkoff.kora.kora.app.annotation.processor.declaration.ComponentDeclaration declaration, DependencyClaim claim) implements ComponentDependency {

        @Override
        public CodeBlock write(ProcessingContext ctx, List<ResolvedComponent> resolvedComponents, Set<Integer> promisedComponents) {
            var dependencies = GraphResolutionHelper.findDependency(ctx, declaration, resolvedComponents, this.claim);
            return CodeBlock.of("g.promiseOf($L.get())", dependencies.component().name());
        }
    }
}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/component/ComponentDependency.kt
package ru.tinkoff.kora.kora.app.ksp.component

import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.KSType
import com.google.devtools.ksp.symbol.KSTypeAlias
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import com.squareup.kotlinpoet.ksp.toTypeParameterResolver
import ru.tinkoff.kora.application.graph.TypeRef
import ru.tinkoff.kora.kora.app.ksp.GraphResolutionHelper
import ru.tinkoff.kora.kora.app.ksp.ProcessingContext
import ru.tinkoff.kora.kora.app.ksp.declaration.ComponentDeclaration
import ru.tinkoff.kora.ksp.common.CommonClassNames

sealed interface ComponentDependency {
    val claim: DependencyClaim

    fun write(ctx: ProcessingContext, resolvedComponents: List<ResolvedComponent>): CodeBlock

    sealed interface SingleDependency : ComponentDependency {
        val component: ResolvedComponent?
    }

    data class TargetDependency(override val claim: DependencyClaim, override val component: ResolvedComponent) : SingleDependency {
        override fun write(ctx: ProcessingContext, resolvedComponents: List<ResolvedComponent>): CodeBlock {
            return CodeBlock.of("it.get(%L)", component.name)
        }
    }

    data class WrappedTargetDependency(override val claim: DependencyClaim, override val component: ResolvedComponent) : SingleDependency {
        override fun write(ctx: ProcessingContext, resolvedComponents: List<ResolvedComponent>): CodeBlock {
            return CodeBlock.of("it.get(%L).value()", component.name)
        }
    }

    data class NullDependency(override val claim: DependencyClaim) : SingleDependency {
        override val component = null

        override fun write(ctx: ProcessingContext, resolvedComponents: List<ResolvedComponent>): CodeBlock {
            return when (claim.claimType) {
                DependencyClaim.DependencyClaimType.NULLABLE_ONE -> CodeBlock.of("null as %T", claim.type.toTypeName().copy(true))
                DependencyClaim.DependencyClaimType.NULLABLE_VALUE_OF -> CodeBlock.of("null as %T", CommonClassNames.valueOf.parameterizedBy(claim.type.toTypeName()).copy(true))
                DependencyClaim.DependencyClaimType.NULLABLE_PROMISE_OF -> CodeBlock.of("null as %T", CommonClassNames.promiseOf.parameterizedBy(claim.type.toTypeName()).copy(true))
                else -> throw IllegalArgumentException(claim.claimType.toString())
            }
        }
    }


    data class ValueOfDependency(override val claim: DependencyClaim, val delegate: SingleDependency) : SingleDependency {
        override val component
            get() = delegate.component

        override fun write(ctx: ProcessingContext, resolvedComponents: List<ResolvedComponent>): CodeBlock {
            if (delegate is NullDependency) {
                return CodeBlock.of("%T.valueOfNull()", CommonClassNames.valueOf)
            }
            val component = delegate.component!!
            if (delegate is WrappedTargetDependency) {
                return CodeBlock.of("it.valueOf(%L).map { it.value() }.map { it as %T }", component.name, claim.type.toTypeName())
            }
            return CodeBlock.of("it.valueOf(%L).map { it as %T }", component.name, claim.type.toTypeName())
        }
    }

    data class PromiseOfDependency(override val claim: DependencyClaim, val delegate: SingleDependency) : SingleDependency {
        override val component
            get() = delegate.component

        override fun write(ctx: ProcessingContext, resolvedComponents: List<ResolvedComponent>): CodeBlock {
            if (delegate is NullDependency) {
                return CodeBlock.of("%T.promiseOfNull()", CommonClassNames.promiseOf)
            }
            val component = delegate.component!!
            if (delegate is WrappedTargetDependency) {
                return CodeBlock.of("it.promiseOf(%L).map { it.value() }.map { it as %T }", component.name, claim.type.toTypeName())
            }
            return CodeBlock.of("it.promiseOf(%L).map { it as %T }", component.name, claim.type.toTypeName())
        }
    }

    data class TypeOfDependency(override val claim: DependencyClaim) : ComponentDependency {
        override fun write(ctx: ProcessingContext, resolvedComponents: List<ResolvedComponent>): CodeBlock {
            return buildTypeRef(claim.type)
        }

        private fun buildTypeRef(typeRef: KSType): CodeBlock {
            val typeParameterResolver = typeRef.declaration.typeParameters.toTypeParameterResolver()
            var declaration = typeRef.declaration
            if (declaration is KSTypeAlias) {
                declaration = declaration.type.resolve().declaration
            }
            if (declaration is KSClassDeclaration) {
                val b = CodeBlock.builder()
                val typeArguments = typeRef.arguments

                if (typeArguments.isEmpty()) {
                    b.add("%T.of(%T::class.java)", TypeRef::class, declaration.toClassName())
                } else {
                    b.add("%T.of(%T::class.java", TypeRef::class, declaration.toClassName())
                    for (typeArgument in typeArguments) {
                        b.add(",\n%L", buildTypeRef(typeArgument.type!!.resolve()))
                    }
                    b.add("\n)")
                }
                return b.build()
            } else {
                return CodeBlock.of("%T.of(%T::class.java)", TypeRef::class, typeRef.toTypeName(typeParameterResolver))
            }
        }
    }

    data class AllOfDependency(override val claim: DependencyClaim) : ComponentDependency {
        override fun write(ctx: ProcessingContext, resolvedComponents: List<ResolvedComponent>): CodeBlock {
            val codeBlock = CodeBlock.builder().add("%T.of(", CommonClassNames.all)
            val dependencies = GraphResolutionHelper.findDependenciesForAllOf(ctx, claim, resolvedComponents)
            for ((i, dependency) in dependencies.withIndex()) {
                if (i == 0) {
                    codeBlock.indent().add("\n")
                }
                codeBlock.add(dependency.write(ctx, resolvedComponents))
                if (i == dependencies.size - 1) {
                    codeBlock.unindent()
                } else {
                    codeBlock.add(", ")
                }
                codeBlock.add("\n")
            }
            return codeBlock.add(")").build()
        }
    }

    data class PromisedProxyParameterDependency(val declaration: ComponentDeclaration, override val claim: DependencyClaim) : ComponentDependency {
        override fun write(ctx: ProcessingContext, resolvedComponents: List<ResolvedComponent>): CodeBlock {
            val dependencies = GraphResolutionHelper.findDependency(ctx, declaration, resolvedComponents, this.claim)
            return CodeBlock.of("it.promiseOf(self.%L)", dependencies!!.component!!.name)
        }

    }

}
